File: BOJ/BOJ/boj 24729 심각한 계단중독입니다.cpp
#include <iostream>

#define MAX_N 100'000

using namespace std;
int N;
int A[MAX_N + 10];
int min_num = MAX_N + 1;
int max_num = 0;
int remain;
bool is_ok(){
    // 먼저 min -> max, max -> min +1
    // if(N % 2 == 1){
    //     return false;
    // }
    
    for(int i = min_num; i <= max_num; i++){
        if(A[i]){
           A[i]--;
           remain--;
        }else{
            return false;
        }
    }
    for(int i = min_num+1; i < max_num; i++){
        if(A[i]){
            A[i]--;
            remain--;
        }else{
            return false;
        }
    }
    for(int i = min_num; i < max_num; i++){
        if(A[i] == 0){
            continue;
        }
        if(A[i] > A[i+1]){
            return false;
        }
        A[i+1] -= A[i];
        remain -= A[i] * 2;
    }
    return remain == 0;
}

int main(){
    cin.tie(0);
    ios::sync_with_stdio(false);
    cin >> N;
    remain = N;
    for(int i = 0; i < N; i++){
        int in;
        cin >> in;
        A[in]++;
        min_num = min(min_num, in);
        max_num = max(max_num, in);
    }

    if(is_ok()){
        cout << 1 << "\n";
    }else{
        cout << -1 << "\n";
    }

    return 0;
}

File: BOJ/BOJ/boj 24729 심각한 계단중독입니다.py
import sys
MAX_NUM = 1000000

input = sys.stdin.readline
sys.setrecursionlimit(MAX_NUM+10)

def solve():
    N = int(input())
    A = [*map(int,input().split())]
    nums = [0 for _ in range(MAX_NUM+1)]
    remain = 0
    min_num = MAX_NUM+1
    for a in A:
        min_num = min(min_num,a)
        nums[a] += 1
        remain += 1
    
    def dfs(x, remain):
        nonlocal min_num,nums
        if remain == 0:
            if x == min_num:
                return True
            else:
                return False
            
        
        for nxt in [x+1, x-1]:
            if nxt < 0 or nxt > MAX_NUM:
                continue
            nums[nxt] -= 1
            if dfs(nxt,remain-1):
                return True
            nums[nxt] += 1
        return False
    
    if dfs(min_num,remain):
        print(1)
    else:
        print(-1)

solve()